detector_test.go

```diff
@@ -585,30 +781,221 @@ func testDetector(t *testing.T, when spec.G, it spec.S) {
 
 			when("with detect error", func() {
 				it("returns a detect error", func() {
-					bpA1 := testmock.NewMockBuildpack(mockCtrl)
-					buildpackStore.EXPECT().Lookup("A", "v1").Return(bpA1, nil)
-					bpA1.EXPECT().ConfigFile().Return(&buildpack.Descriptor{API: "0.3"})
+					bpA1 := testmock.NewMockBuildModule(mockCtrl)
+					dirStore.EXPECT().LookupBp("A", "v1").Return(bpA1, nil)
+					bpA1.EXPECT().ConfigFile().Return(&buildpack.Descriptor{API: "0.3", Buildpack: buildpack.Info{ID: "A", Version: "v1"}})
 					bpA1.EXPECT().Detect(gomock.Any(), gomock.Any())
 
-					group := []buildpack.GroupBuildpack{
+					group := []buildpack.GroupElement{
 						{ID: "A", Version: "v1", API: "0.3"},
 					}
 					resolver.EXPECT().Resolve(group, detector.Runs).Return(
-						[]buildpack.GroupBuildpack{},
+						[]buildpack.GroupElement{},
 						[]platform.BuildPlanEntry{},
 						lifecycle.ErrFailedDetection,
 					)
 
-					_, _, err := detector.Detect(buildpack.Order{{Group: group}})
+					detector.Order = buildpack.Order{{Group: group}}
+					_, _, err := detector.Detect()
 					if err, ok := err.(*buildpack.Error); !ok || err.Type != buildpack.ErrTypeFailedDetection {
 						t.Fatalf("Unexpected error:\n%s\n", err)
 					}
 				})
 			})
 		})
+
+		when("there are extensions", func() {
+			it("selects the first passing group", func() {
+				// This test doesn't use gomock.InOrder() because each call to Detect() happens in a go func.
```

Comment:
Oof, these tests need a lot of gymnastics to set up. Probably doesn't need to be part of this, but maybe worth an issue to revisit to make them more manageable? (Or maybe there is enough inherent complexity that this is the way?)

Reply:
Made an issue: https://github.com/buildpacks/lifecycle/issues/866